src/ring_theory/localization.lean

```diff
@@ -891,7 +892,67 @@ lemma map_mem_non_zero_divisors {B : Type*} [integral_domain B] {g : A →+* B}
 λ z hz, eq_zero_of_ne_zero_of_mul_eq_zero
   (map_ne_zero_of_mem_non_zero_divisors hg) hz
 
-variables (K : Type*)
+variables {K}
+
+lemma le_non_zero_divisors_of_domain [integral_domain K] {M : submonoid K} (hM : ↑0 ∉ M)
+  : M ≤ non_zero_divisors K :=
+λ x hx y hy, or.rec_on (eq_zero_or_eq_zero_of_mul_eq_zero hy)
+  (λ h, h) (λ h, absurd (h ▸ hx : (0 : K) ∈ M) hM)
+
+namespace localization_map
+
+lemma to_map_eq_zero_iff (f : localization_map M S) {x : R}
+  (hM : M ≤ non_zero_divisors R) : x = 0 ↔ f.to_map x = 0 :=
+begin
+  rw ← f.to_map.map_zero,
+  split; intro h,
+  { rw h },
+  { cases f.eq_iff_exists.mp h with c hc,
+    rw zero_mul at hc,
+    exact hM c.2 x hc }
+end
+
+lemma injective (f : localization_map M S)
+  (hM : M ≤ non_zero_divisors R) : injective f.to_map :=
+begin
+  rw ring_hom.injective_iff f.to_map,
+  intros a ha,
+  rw [← f.to_map.map_zero, f.eq_iff_exists] at ha,
+  cases ha with c hc,
+  rw zero_mul at hc,
+  exact hM c.2 a hc,
+end
+
+protected lemma map_ne_zero_of_mem_non_zero_divisors {M : submonoid A} (f : localization_map M S)
+  (hM : M ≤ non_zero_divisors A) (x : non_zero_divisors A) : f.to_map x ≠ 0 :=
+map_ne_zero_of_mem_non_zero_divisors (f.injective hM)
+
+/-- A `comm_ring` `S` which is the localization of an integral domain `R` at a subset of `R - {0}`
+is an integral domain. -/
+def to_integral_domain {M : submonoid A} (f : localization_map M S)
+  (hM : M ≤ non_zero_divisors A) : integral_domain S :=
+{ eq_zero_or_eq_zero_of_mul_eq_zero :=
+    begin
+      intros z w h,
+      cases f.surj z with x hx,
+      cases f.surj w with y hy,
+      have : z * w * f.to_map y.2 * f.to_map x.2 = f.to_map x.1 * f.to_map y.1,
+      by rw [mul_assoc z, hy, ←hx]; ac_refl,
+      rw [h, zero_mul, zero_mul, ← f.to_map.map_mul] at this,
+      cases eq_zero_or_eq_zero_of_mul_eq_zero ((to_map_eq_zero_iff f hM).mpr this.symm) with H H,
+      { exact or.inl (f.eq_zero_of_fst_eq_zero hx H) },
+      { exact or.inr (f.eq_zero_of_fst_eq_zero hy H) },
+    end,
+  exists_pair_ne := ⟨f.to_map 0, f.to_map 1, λ h, zero_ne_one (f.injective hM h)⟩,
+  ..(infer_instance : comm_ring S) }
+
+@[priority 100]
+instance integral_domain_localization {M : submonoid A} (hM : M ≤ non_zero_divisors A)
```

Comment:
```suggestion
def integral_domain_localization {M : submonoid A} (hM : M ≤ non_zero_divisors A)
```
This instance will not trigger, because typeclass inference can now no longer find `hM`. But you can state a dedicated instance for `fraction_map`. (I guess that instance is already there, but the proof is now a 1-liner.